File: code_run_AngleCam/run_AngleCam.R

### ----------------------------------------------------------------------------------------------
### ---- AngleCam --------------------------------------------------------------------------------
### ----------------------------------------------------------------------------------------------

# This is the alpha version of AngleCam (2023-06-05)
# AngleCam estimates leaf angle distributions from conventional RGB photographs
# for details see: https://github.com/tejakattenborn/AngleCAM
# AngleCam can be applied on horizontally acquired imagery with a minumum image size of 600x600 px.
# Input imagery can be of different size; the pipeline will resample them on the fly to 600x600 px.
# Find below two examples for visualizing estimates as leaf angle distribution or average leaf angle

# -------TensorFlow------------------------------------------------------------------------------

# Sketch and some info on installing TensorFlow (sometimes tricky):
# devtools::install_github("rstudio/tensorflow")
# install_tensorflow(version = "2.11.0")
# > official github repository: https://github.com/rstudio/tensorflow
# > General info on installing Tensorflow: https://tensorflow.rstudio.com/installation/
# > running TensorFlow with GPU support is recommended but not mandatory (CPU also works fine, but slower)
# > Hints for TensorFlow with GPU support: https://tensorflow.rstudio.com/installation/gpu/local_gpu/
# > hello <- tf$constant("Hello"); print(hello) # check TensorFlow installation

# ------Dependencies--------------------------------------------------------------------------------

library(reticulate)
reticulate::use_condaenv(condaenv = "tfr", required = TRUE)

library(tensorflow)
library(keras)
library(dplyr)
library(tfdatasets)
require(purrr)

setwd("INSERT DIR")

### GPU settings (only needed if tensorflow is run on GPU, recommended but not mandatory; CPU also works)
gpu_no = 1 # GPU 1 or 2? (in case of multiple GPUs)
gpu_no = tf$config$list_physical_devices()[c(1, gpu_no +1)]
tf$config$set_visible_devices(gpu_no)

### compile the model and initialize the tfdatasets pipeline
source("cnn_reqs.R")
# load model weights
model <- load_model_hdf5("AngleCam_efficientnet_V2L_14-03-2023.hdf5", compile = FALSE)

# -------AngleCam prediction----------------------------------------------------------------------

### load image data and create tfdataset (the imagery will not be loaded to RAM but processed on the fly; there are, hence, no limits regarding dataset size)
imagepaths <- list.files(path = "example_dataset_timeseries_tilia_cordata", full.names = T, pattern = ".png", recursive = F); paste0(length(imagepaths), " images found.")
imgdataset <- create_dataset(tibble(img = imagepaths),  batch = 1, epochs = 1, shuffle = F, train = F, aug_geo = F, aug_rad = F)

### apply the model to the imagery and write output to file
pred_imgdataset <- model %>% predict(imgdataset)
pred_imgdataset <- pred_imgdataset/10
rownames(pred_imgdataset) <- basename(imagepaths)
write.csv(pred_imgdataset, file = "AngleCam_predictions_example_dataset_timeseries_tilia_cordata.csv", row.names = T)

#--------simple visulization--------------------------------------------------------------------

### plot leaf angle distributions over time
for(i in 1:nrow(pred_imgdataset)){
  if (i%%10 == 0) {
    #lines(seq(0,90, length.out = 43), pred_imgdataset[i,])
    plot(seq(0,90, length.out = 43), pred_imgdataset[i,], ylim=c(0,0.07), xlim=c(0,90), type="l", main = substr(rownames(pred_imgdataset)[i], 15, 33), ylab = "density", xlab = "leaf angle [deg]")
    Sys.sleep(0.6)
  }
}

### plot average leaf angles over time
# derive average leaf angle from leaf angle distributions
mean_angle = function(pred){sum(pred*seq(0,90, length.out = ncol(pred_imgdataset)))}
avg_angle <- apply(pred_imgdataset, 1, mean_angle)
# derive date from filename
dates = as.POSIXct(substr(rownames(pred_imgdataset), 15, 33), tz = "Europe/Berlin", format = "%Y-%m-%d_%H-%M-%OS")
# plot leaf angles for one day
plot(dates, avg_angle, ylab = "Average leaf angle [deg]", xlab = "time (CEST)")

File: code_run_AngleCam/run_AngleCam.py
# Code Description:

# In this code, the parameters and their explanations are as follows:

# - `directory`: The directory path where the image data is located. It can be set to the desired directory path.
# - `model_path`: The file path to the model weights file (.hdf5) that will be loaded for inference.
# - `imagepaths`: A list of image file paths to be processed. These files should be located in the "wildcam12_test" folder within the directory.

# The code performs the following steps:

# 1. Loads the model weights using `keras.models.load_model`.
# 2. Defines a function `get_image_taken_datetime` to extract the image taken date and time from the metadata using PIL.
# 3. Loads the image data from the specified folder using `glob.glob` and stores the image file paths in `imagepaths`.
# 4. Processes and resizes the images using OpenCV, scaling the pixel values between 0 and 1.
# 5. Applies the loaded model to the processed images and saves the predictions to a CSV file.
# 6. Extracts the date and time metadata from the images using the `get_image_taken_datetime` function.
# 7. Calculates the average angle from the prediction dataset.
# 8. Creates a DataFrame (`df`) containing the extracted dates, average angles, and file names.
# 9. Saves the DataFrame to a CSV file.
# 10. Plots the average leaf angles over time using Matplotlib and saves the plot as an image.

# Please make sure to modify the relevant parameters and file paths according to your specific setup.

import tensorflow as tf
import keras
import pandas as pd
import glob
import os
os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"
import cv2
import matplotlib.pyplot as plt
import numpy as np
#from PIL import Image
#from PIL.ExifTags import TAGS

# Set the directory where the image data are located
directory = os.getcwd()
print("Working directory: " + str(directory))

# Load the model weights
model_path = "G:/My Drive/1_coding/anglecam/1_AngleCam_application/v2/AngleCam_efficientnet_V2L_14-03-2023.hdf5"
model = keras.models.load_model(model_path)

# Check if the model is loaded correctly
if model:
    print("Model loaded successfully.")
else:
    print("Failed to load the model.")


# Set GPU settings if running TensorFlow on GPU
gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    tf.config.experimental.set_memory_growth(gpus[0], True)

# Load the image data
imagepaths = glob.glob(os.path.join(directory, "example_dataset_timeseries_tilia_cordata", "*.png"))
print("Taking following imagepath: " + str(imagepaths))
print(f"{len(imagepaths)} images found.")

# Process and resize the images
processed_images = []
for imagepath in imagepaths:
    image = cv2.imread(imagepath)
    resized_image = cv2.resize(image, (600, 600))
    processed_image = resized_image.astype(np.float32) / 255.0
    processed_images.append(processed_image)


imgdataset = tf.data.Dataset.from_tensor_slices(processed_images)

# Apply the model to the imagery and write output to file
pred_imgdataset = model.predict(imgdataset.batch(1))
pred_imgdataset = pred_imgdataset / 10
pred_imgdataset = pd.DataFrame(pred_imgdataset, index=[os.path.basename(path) for path in imagepaths])
pred_imgdataset.to_csv(os.path.join(directory, "AngleCam_prediction_table.csv"))

print("Column titles in pred_imgdataset:")
print(pred_imgdataset.columns)

# Extract the date and time metadata from the images
file_names = []
for imagepath in imagepaths:
    file_names.append(os.path.basename(imagepath))

dates = []
for filename in file_names:
    dates.append(pd.to_datetime(filename[14:33], format="%Y-%m-%d_%H-%M-%S", utc=True))


# Convert filenames to datetime objects
dates = [datetime.datetime.strptime(filename[-19:-9], "%Y-%m-%d") for filename in filenames]

def mean_angle(row):
    pred = row.iloc[1:]
    length = len(pred)
    angles = np.linspace(0, 90, num=length)
    result = np.sum(pred * angles)
    return result

avg_angle = pred_imgdataset.apply(mean_angle, axis=1)

print("Average angle: " + str(avg_angle))
data = {"dates": dates, "avg_angle": avg_angle, "file_name": file_names}
df = pd.DataFrame(data)

# Save the data to a CSV file
df.to_csv(os.path.join(directory, "leaf_angle_data.csv"), index=False)

# Plot average leaf angles over time
plt.scatter(dates, avg_angle)
plt.xlabel("Time (CEST)")
plt.ylabel("Average leaf angle [deg]")
plt.title("Average Leaf Angle over Time")
plt.savefig("leaf_angle_plot.png")

